tiny_cnn/network.h

```diff
@@ -672,10 +672,18 @@ class network
         for_i(num_threads, [&](int i) {
             int start_index = i * data_per_thread;
             int end_index = std::min(batch_size, start_index + data_per_thread);
-
-            // loop over data points in this batch assigned to thread i
-            for (int j = start_index; j < end_index; ++j)
-                bprop<E>(fprop(in[j], i), t[j], i, t_cost ? &(t_cost[j]) : nullptr);
+            
+            if (start_index < end_index) { // any data for this thread to process?
+                // process data points in this batch assigned to thread i
+                std::vector<tensor_t> in_batch(&in[start_index], &in[end_index]);
+                std::vector<tensor_t> t_batch(&t[start_index], &t[end_index]);
+                std::vector<tensor_t> t_cost_batch
+                    = t_cost
+                    ? std::vector<tensor_t>(&t_cost[start_index], &t_cost[end_index])
+                    : std::vector<tensor_t>();
+
+                bprop<E>(fprop(in_batch, i), t_batch, i, t_cost_batch);
```

Comment:
This `bprop`s only those samples from the minibatch that have been assigned to this thread. I guess this will need to be changed when implementing Batch Normalization, so that the whole minibatch goes to the layer as a single object. The parallelism can then be handled inside each layer. This may be a good thing, because the optimal parallelism strategy may be different for each layer type (at least in theory).